include/RAJA/policy/cuda/nested.hpp

```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
 };
 
 
-struct cuda_collapse_exec{};
+namespace internal {
 
-template <typename FT0, typename FT1>
-struct Executor<Collapse<cuda_collapse_exec, FT0, FT1>> {
 
-  // TODO, check that FT0 and FT1 are cuda policies
+
+/*!
+ * \brief Launcher that uses execution policies to map blockIdx and threadIdx to
+ * map
+ * to N-argument function
+ */
+template <typename BODY, typename... CARGS>
+__global__ void cudaLauncher(BODY loop_body, CARGS... cargs)
+{
+  // force reduction object copy constructors and destructors to run
+  auto body = loop_body;
+
+  // Invoke the wrapped body.
+  // The wrapper will take care of computing indices, and deciding if the
+  // given block+thread is in-bounds, and invoking the users loop body
+  body();
+}
+
+
+} // namespace internal
+
+
+template<bool Async = false>
+struct cuda_collapse_exec : public cuda_exec<0, Async>
+{};
+
+
+template<typename ... FOR>
+using CudaCollapse = Collapse<cuda_collapse_exec<false>, FOR ...>;
+
+template<typename ... FOR>
+using CudaCollapseAsync = Collapse<cuda_collapse_exec<true>, FOR ...>;
+
+
+// TODO, check that FT... are cuda policies
```

Comment:
It may be worth just doing this, now that the check is actually pretty easy.  See [here](https://github.com/LLNL/RAJA/blob/develop/include/RAJA/pattern/forallN.hpp#L61).